documentation/sphinx/source/configuration.rst
.. |multiplicative-suffixes| replace:: Sizes must be specified as a number of bytes followed by one of the multiplicative suffixes B=1, KB=10\ :sup:`3`, KiB=2\ :sup:`10`, MB=10\ :sup:`6`, MiB=2\ :sup:`20`, GB=10\ :sup:`9`, GiB=2\ :sup:`30`, TB=10\ :sup:`12`, or TiB=2\ :sup:`40`.

#############
Configuration
#############

.. include:: guide-common.rst.inc

This document contains *reference* information for configuring a new FoundationDB cluster. We recommend that you read this document before setting up a cluster for performance testing or production use. For *step-by-step instructions* to follow when setting up a cluster, see :doc:`building-cluster`.

.. note:: In FoundationDB, a "cluster" refers to one or more FoundationDB processes spread across one or more physical machines that together host a FoundationDB database.

To plan an externally accessible cluster, you need to understand some basic aspects of the system. You can start by reviewing the :ref:`system requirements <system-requirements>`, then how to :ref:`choose <configuration-choosing-coordination-servers>` and :ref:`change coordination servers <configuration-changing-coordination-servers>`. Next, you should look at the :ref:`configuration file <foundationdb-conf>`, which controls most other aspects of the system. Then, you should understand how to :ref:`choose a redundancy mode <configuration-choosing-redundancy-mode>` and :ref:`configure the storage subsystem <configuration-configuring-storage-subsystem>`. Finally, there are some guidelines for setting :ref:`process class configurations <guidelines-process-class-config>`.

.. _system-requirements:

System requirements
===================

* One of the following 64-bit operating systems:

  * A supported Linux distribution:
  
    * RHEL/CentOS 6.x and 7.x
    * Ubuntu 12.04 or later (but see :ref:`Platform Issues for Ubuntu 12.x <platform-ubuntu-12>`)

  * Or, an unsupported Linux distribution with:

    * Kernel version between 2.6.33 and 3.0.x (inclusive) or 3.7 or greater
    * Preferably .deb or .rpm package support
  
  * Or, macOS 10.7 or later
  
  .. warning:: The macOS and Windows versions of the FoundationDB server are intended for use on locally accessible development machines only. Other uses are not supported.
  
* 4GB **ECC** RAM (per fdbserver process)
* Storage

  * SSDs are required when storing data sets larger than memory (using the ``ssd`` storage engine).
  * HDDs are OK when storing data sets smaller than memory (using the ``memory`` storage engine).
  * For more information, see :ref:`configuration-configuring-storage-subsystem`.

For a description of issues on particular platforms that affect the operation of FoundationDB, see :doc:`platforms`.

.. _configuration-choosing-coordination-servers:

Choosing coordination servers
=============================

FoundationDB uses a set of *coordination servers* (or *coordinators* for short) to maximize the fault tolerance (and, in particular, the availability) of the cluster. The coordinators work by communicating and storing a small amount of shared state. If one or more machines are down or unable to communicate with the network, the cluster may become partitioned. In that event, FoundationDB selects the partition in which a majority of coordinators are reachable as the one that will remain available.

Any FoundationDB process can be used as a coordinator for any set of clusters. The performance impact of acting as a coordinator is negligible. The coordinators aren't involved at all in committing transactions.

Administrators should choose the number and physical location of coordinators to maximize fault tolerance. Most configurations should follow these guidelines:

* Choose an odd number of coordinators.
* Use enough coordinators to complement the :ref:`redundancy mode <configuration-choosing-redundancy-mode>` of the cluster, often 3 or 5.
* Place coordinators in different racks, circuits, or datacenters with independence of failure.
* It is OK to place coordinators in distant datacenters; in normal operation the latency to a coordinator does not affect system latency.

The set of coordinators is stored on each client and server in the :ref:`cluster file <foundationdb-cluster-file>`.

.. _configuration-changing-coordination-servers:

Changing coordination servers
=============================

It is sometimes necessary to change the set of coordinators servers. You may want to do so because of changing network conditions, machine failures, or just planning adjustments. You can change coordinators using an automated ``fdbcli`` command. FoundationDB will maintain ACID guarantees during the changes.

You can change coordinators when the following conditions are met:

* a majority of the current coordinators are available;
* all of the new coordinators are available; and
* client and server cluster files and their parent directories are writable.

``fdbcli`` supports a ``coordinators`` command to specify the new list of coordinators::

    user@host$ fdbcli
    Using cluster file `/etc/foundationdb/fdb.cluster'.

    The database is available.

    Welcome to the fdbcli. For help, type `help'.
    fdb> coordinators 10.0.4.1:4500 10.0.4.2:4500 10.0.4.3:4500
    Coordinators changed

After running this command, you can check that it completed successfully by using the ``status details`` command::

    fdb> status details

    Configuration:
      Redundancy mode        - triple
      Storage engine         - ssd
      Coordinators           - 3

    Cluster:
      FoundationDB processes - 3
      Machines               - 3
      Memory availability    - 4.1 GB per process on machine with least available
      Fault Tolerance        - 0 machines
      Server time            - Thu Mar 15 14:41:34 2018

    Data:
      Replication health     - Healthy
      Moving data            - 0.000 GB
      Sum of key-value sizes - 8 MB
      Disk space used        - 103 MB

    Operating space:
      Storage server         - 1.0 GB free on most full server
      Transaction log        - 1.0 GB free on most full server

    Workload:
      Read rate              - 2 Hz
      Write rate             - 0 Hz
      Transactions started   - 2 Hz
      Transactions committed - 0 Hz
      Conflict rate          - 0 Hz

    Backup and DR:
      Running backups        - 0
      Running DRs            - 0

    Process performance details:
      10.0.4.1:4500       ( 3% cpu;  2% machine; 0.004 Gbps;  0% disk; 2.5 GB / 4.1 GB RAM  )
      10.0.4.2:4500       ( 1% cpu;  2% machine; 0.004 Gbps;  0% disk; 2.5 GB / 4.1 GB RAM  )
      10.0.4.3:4500       ( 1% cpu;  2% machine; 0.004 Gbps;  0% disk; 2.5 GB / 4.1 GB RAM  )

    Coordination servers:
      10.0.4.1:4500
      10.0.4.2:4500
      10.0.4.3:4500

    Client time: Thu Mar 15 14:41:34 2018
    
The list of coordinators verifies that the coordinator change succeeded. A few things might cause this process to not go smoothly:

* If any of the new coordination servers fail before you run the ``coordinators`` command, the change will not occur, and the database will continue to use the old coordinators.
* If a majority of the new coordination servers fail during or after the change, the database will not be available until a majority of them are available again.
* If a majority of the old coordination servers fail before the change is completed, the database will be unavailable until a majority of them are available again. Consequently, the ``coordinators`` command cannot be used to repair a database which is unavailable because its coordinators are down.

Once the change is complete, database servers and clients need to communicate with the old coordinators in order to update their cluster file to point to the new coordinators. (Each database server and client will physically re-write their cluster file to reference the new coordinators.) In most cases this process occurs automatically within a matter of seconds.

If some servers or clients are unable to write to their cluster file or are disconnected during the change of coordinators (and too many of the old coordinators become unavailable before they come back up), an administrator will need to manually copy the new cluster file to each affected machine and restart the database server or client, as if adding a new server or client to the cluster.

The ``coordinators`` command also supports a convenience option, ``coordinators auto``, that automatically selects a set of coordination servers appropriate for the redundancy mode::

    user@host1$ fdbcli
    Using cluster file `/etc/foundationdb/fdb.cluster'.

    The database is available.

    Welcome to the fdbcli. For help, type `help'.
    fdb> coordinators auto
    Coordinators changed

``coordinators auto`` will not make any changes if the current coordinators are all available and support the current redundancy level.

.. note:: |coordinators-auto|

.. _configuration-setting-cluster-description:

Changing the cluster description
================================

A cluster is named by the ``description`` field recorded in the ``fdb.cluster`` file (see :ref:`cluster-file-format`). For convenience of reference, you may want to change the ``description`` if you operate more than one cluster on the same machines. (Each cluster must be uniquely identified by the combination of ``description`` and ``ID``.)

You can change the ``description`` using the ``coordinators`` command within ``fdbcli``::

    fdb> coordinators description=cluster_b
    Coordination state changed

You can also combine a change of ``description`` with :ref:`changing coordinators <configuration-changing-coordination-servers>`, whether by listing the coordinators or with ``coordinators auto``::

    fdb> coordinators auto description=cluster_c
    Coordination state changed

.. _foundationdb-conf:

The configuration file
======================

The ``fdbserver`` server process is run and monitored on each server by the ``fdbmonitor`` :ref:`daemon <administration_fdbmonitor>`. ``fdbmonitor`` and ``fdbserver`` itself are controlled by the ``foundationdb.conf`` file located at:

* ``/etc/foundationdb/foundationdb.conf`` on Linux
* ``/usr/local/etc/foundationdb/foundationdb.conf`` on macOS

The ``foundationdb.conf`` file contains several sections, detailed below. Note that the presence of individual ``[fdbserver.<ID>]`` sections actually cause ``fdbserver`` processes to be run.

.. note:: |conf-file-change-detection|

.. warning:: Do not attempt to stop FoundationDB services by removing the configuration file. Removing the file will not stop the services; it will merely remove your ability to control them in the manner supported by FoundationDB. During normal operation, services can be stopped by commenting out or removing the relevant sections of the configuration file. You can also disable a service at the operating system level or by removing the software.

``[fdbmonitor]`` section
------------------------

.. code-block:: ini

    ## foundationdb.conf 
    ##
    ## Configuration file for FoundationDB server processes 

    [fdbmonitor]
    user = foundationdb
    group = foundationdb

Contains basic configuration parameters of the ``fdbmonitor`` process. ``user`` and ``group`` are used on Linux systems to control the privilege level of child processes.

``[general]`` section
-----------------------

.. code-block:: ini

    [general]
    cluster-file = /etc/foundationdb/fdb.cluster
    restart-delay = 60
    ## restart-backoff and restart-delay-reset-interval default to the value that is used for restart-delay
    # initial-restart-delay = 0
    # restart-backoff = 60.0
    # restart-delay-reset-interval = 60
    # envvars = FOO=BAR BAZ=FOO
    # delete-envvars =
    # kill-on-configuration-change = true
    # disable-lifecycle-logging = false

Contains settings applicable to all processes (e.g. fdbserver, backup_agent).

* ``cluster-file``: Specifies the location of the cluster file. This file and the directory that contains it must be writable by all processes (i.e. by the user or group set in the ``[fdbmonitor]`` section).
* ``envvars``: A space separated list of environment variables to be added to child processes' environment variables. Every environment variable in the list must be of the form ``A=B`` where ``A`` and ``B`` are key and value names respectively.
* ``delete-envvars``: A space separated list of environment variables to remove from the environments of child processes. This can be used if the ``fdbmonitor`` process needs to be run with environment variables that are undesired in its children. This takes precedence over envvars field.
* ``kill-on-configuration-change``: If ``true``, affected processes will be restarted whenever the configuration file changes. Defaults to ``true``.
* ``disable-lifecycle-logging``: If ``true``, ``fdbmonitor`` will not write log events when processes start or terminate. Defaults to ``false``.

.. _configuration-restarting:

The ``[general]`` section also contains some parameters to control how processes are restarted when they die. ``fdbmonitor`` uses backoff logic to prevent a process that dies repeatedly from cycling too quickly, and it also introduces up to +/-10% random jitter into the delay to avoid multiple processes all restarting simultaneously. ``fdbmonitor`` tracks separate backoff state for each process, so the restarting of one process will have no effect on the backoff behavior of another.

* ``restart-delay``: The maximum number of seconds (subject to jitter) that fdbmonitor will delay before restarting a failed process.
* ``initial-restart-delay``: The number of seconds ``fdbmonitor`` waits to restart a process the first time it dies. Defaults to 0 (i.e. the process gets restarted immediately). 
* ``restart-backoff``: Controls how quickly ``fdbmonitor`` backs off when a process dies repeatedly. The previous delay (or 1, if the previous delay is 0) is multiplied by ``restart-backoff`` to get the next delay, maxing out at the value of ``restart-delay``. Defaults to the value of ``restart-delay``, meaning that the second and subsequent failures will all delay ``restart-delay`` between restarts.
* ``restart-delay-reset-interval``: The number of seconds a process must be running before resetting the backoff back to the value of ``initial-restart-delay``. Defaults to the value of ``restart-delay``.

 These ``restart-`` parameters are not applicable to the ``fdbmonitor`` process itself. See :ref:`Configuring autorestart of fdbmonitor <configuration-restart-fdbmonitor>` for details.

As an example, let's say the following parameters have been set:

.. code-block:: ini

    restart-delay = 60
    initial-restart-delay = 0
    restart-backoff = 2.0
    restart-delay-reset-interval = 180

The progression of delays for a process that fails repeatedly would be ``0, 2, 4, 8, 16, 32, 60, 60, ...``, each subject to a 10% random jitter. After the process stays alive for 180 seconds, the backoff would reset and the next failure would restart the process immediately.

Using the default parameters, a process will restart immediately if it fails and then delay ``restart-delay`` seconds if it fails again within ``restart-delay`` seconds. 

.. _foundationdb-conf-fdbserver:

``[fdbserver]`` section
-----------------------

.. code-block:: ini
   
    ## Default parameters for individual fdbserver processes
    [fdbserver]
    command = /usr/sbin/fdbserver
    public-address = auto:$ID
    listen-address = public
    datadir = /var/lib/foundationdb/data/$ID
    logdir = /var/log/foundationdb
    # logsize = 10MiB
    # maxlogssize = 100MiB
    # class = 
    # memory = 8GiB
    # memory-vsize =
    # storage-memory = 1GiB
    # cache-memory = 2GiB
    # locality-machineid =
    # locality-zoneid =
    # locality-data-hall =
    # locality-dcid =
    # io-trust-seconds = 20

Contains default parameters for all fdbserver processes on this machine. These same options can be overridden for individual processes in their respective ``[fdbserver.<ID>]`` sections. In this section, the ID of the individual fdbserver can be substituted by using the ``$ID`` variable in the value. For example, ``public-address = auto:$ID`` makes each fdbserver listen on a port equal to its ID.

.. note:: |multiplicative-suffixes|
.. note:: In general locality id's are used to specify the location of processes which in turn is used to determine fault and replication domains.

* ``command``: The location of the ``fdbserver`` binary.
* ``public-address``: The publicly visible IP:Port of the process. If ``auto``, the address will be the one used to communicate with the coordination servers.
* ``listen-address``: The IP:Port that the server socket should bind to. If ``public``, it will be the same as the public-address.
* ``datadir``: A writable directory (by root or by the user set in the [fdbmonitor] section) where persistent data files will be stored.
* ``tlog-spill-datadir``: An optional writable directory where transaction logs store spilled data. Transaction log disk queues remain in ``datadir``. If unset, transaction logs store spilled data in ``datadir`` as well. The directory must be unique for each ``fdbserver`` process. Existing transaction log spill files in ``datadir`` remain readable after this setting is changed. To move existing spill files to the new disk as well, move the ``log-*`` and ``log2-*`` files while the process is stopped. On Linux, when ``data-filesystem`` validation is enabled and this directory is on a separate mount, configure ``tlog-spill-filesystem`` with the additional mount path.
* ``logdir``: A writable directory (by root or by the user set in the [fdbmonitor] section) where FoundationDB will store log files.
* ``logsize``: Roll over to a new log file after the current log file reaches the specified size. The default value is 10MiB.
* ``maxlogssize``: Delete the oldest log file when the total size of all log files exceeds the specified size. If set to 0B, old log files will not be deleted. The default value is 100MiB.
* ``class``: Process class specifying the roles that will be taken in the cluster. Recommended options are ``storage``, ``transaction``, ``stateless``. See :ref:`guidelines-process-class-config` for process class config recommendations.
* ``memory``: Maximum resident memory used by the process. The default value is 8GiB. When specified without a unit, MiB is assumed. Setting to 0 means unlimited. This parameter does not change the memory allocation of the program. Rather, it sets a hard limit beyond which the process will kill itself and be restarted. The default value of 8GiB is double the intended memory usage in the default configuration (providing an emergency buffer to deal with memory leaks or similar problems). It is *not* recommended to decrease the value of this parameter below its default value. It may be *increased* if you wish to allocate a very large amount of storage engine memory or cache. In particular, when the ``storage-memory``  or ``cache-memory`` parameters are increased, the ``memory`` parameter should be increased by an equal amount.

contrib/TestHarness2/test_harness/config.py
from __future__ import annotations

import argparse
import collections
import copy
import os
import random
from enum import Enum
from pathlib import Path
from typing import List, Any, OrderedDict, Dict


class BuggifyOptionValue(Enum):
    ON = 1
    OFF = 2
    RANDOM = 3


class BuggifyOption:
    def __init__(self, val: str | None = None):
        self.value = BuggifyOptionValue.RANDOM
        if val is not None:
            v = val.lower()
            if v in ["on", "1", "true"]:
                self.value = BuggifyOptionValue.ON
            elif v in ["off", "0", "false"]:
                self.value = BuggifyOptionValue.OFF
            elif v in ["random", "rnd", "r"]:
                pass
            else:
                assert False, "Invalid value {} -- use true, false, or random".format(v)


class ConfigValue:
    def __init__(self, name: str, **kwargs):
        self.name = name
        self.value = None
        self.kwargs = kwargs
        if "default" in self.kwargs:
            self.value = self.kwargs["default"]

    def get_arg_name(self) -> str:
        if "long_name" in self.kwargs:
            return self.kwargs["long_name"]
        else:
            return self.name

    def add_to_args(self, parser: argparse.ArgumentParser):
        kwargs = copy.copy(self.kwargs)
        long_name = self.name
        short_name = None
        if "long_name" in kwargs:
            long_name = kwargs["long_name"]
            del kwargs["long_name"]
        if "short_name" in kwargs:
            short_name = kwargs["short_name"]
            del kwargs["short_name"]
        if "action" in kwargs and kwargs["action"] in ["store_true", "store_false"]:
            del kwargs["type"]
        long_name = long_name.replace("_", "-")
        if short_name is None:
            # line below is useful for debugging
            # print('add_argument(\'--{}\', [{{{}}}])'.format(long_name, ', '.join(['\'{}\': \'{}\''.format(k, v)
            #                                                                       for k, v in kwargs.items()])))
            parser.add_argument("--{}".format(long_name), **kwargs)
        else:
            # line below is useful for debugging
            # print('add_argument(\'-{}\', \'--{}\', [{{{}}}])'.format(short_name, long_name,
            #                                                          ', '.join(['\'{}\': \'{}\''.format(k, v)
            #                                                                     for k, v in kwargs.items()])))
            parser.add_argument(
                "-{}".format(short_name), "--{}".format(long_name), **kwargs
            )

    def get_value(self, args: argparse.Namespace) -> tuple[str, Any]:
        return self.name, args.__getattribute__(self.get_arg_name())


class Config:
    """
    This is the central configuration class for test harness. The values in this class are exposed globally through
    a global variable test_harness.config.config. This class provides some "magic" to keep test harness flexible.
    Each parameter can further be configured using an `_args` member variable which is expected to be a dictionary.
    * The value of any variable can be set through the command line. For a variable named `variable_name` we will
      by default create a new command line option `--variable-name` (`_` is automatically changed to `-`). This
      default can be changed by setting the `'long_name'` property in the `_arg` dict.
    * In addition the user can also optionally set a short-name. This can be achieved by setting the `'short_name'`
      property in the `_arg` dictionary.
    * All additional properties in `_args` are passed to `argparse.add_argument`.
    * If the default of a variable is `None` the user should explicitly set the `'type'` property to an appropriate
      type.
    * In addition to command line flags, all configuration options can also be controlled through environment variables.
      By default, `variable-name` can be changed by setting the environment variable `TH_VARIABLE_NAME`. This default
      can be changed by setting the `'env_name'` property.
    * Test harness comes with multiple executables. Each of these should use the config facility. For this,
      `Config.build_arguments` should be called first with the `argparse` parser. Then `Config.extract_args` needs
      to be called with the result of `argparse.ArgumentParser.parse_args`. A sample example could look like this:
      ```
      parser = argparse.ArgumentParser('TestHarness', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
      config.build_arguments(parser)
      args = parser.parse_args()
      config.extract_args(args)
      ```
    * Changing the default value for all executables might not always be desirable. If it should be only changed for
      one executable Config.change_default should be used.
    """

    def __init__(self):
        self.random = random.Random()
        self.cluster_file: str | None = None
        self.cluster_file_args = {
            "short_name": "C",
            "type": str,
            "help": "Path to fdb cluster file",
            "required": False,
            "env_name": "JOSHUA_CLUSTER_FILE",
        }
        self.joshua_dir: str | None = None
        self.joshua_dir_args = {
            "type": str,
            "help": "Where to write FDB data to",
            "required": False,
            "env_name": "JOSHUA_APP_DIR",
        }
        self.joshua_output_dir: Path | None = None
        self.joshua_output_dir_args = {
            "type": Path,
            "default": None,
            "help": "Directory for TestHarness2 to store output files.",
        }
        self.stats: str | None = None
        self.stats_args = {
            "type": str,
            "help": "A base64 encoded list of statistics (used to reproduce runs)",
            "required": False,
        }
        self.random_seed: int | None = None
        self.random_seed_args = {
            "type": int,
            "help": "Force given seed given to fdbserver -- mostly useful for debugging",
            "required": False,
        }
        self.kill_seconds: int = 30 * 60
        self.kill_seconds_args = {"help": "Timeout for individual test"}
        self.buggify_on_ratio: float = 0.8
        self.buggify_on_ratio_args = {"help": "Probability that buggify is turned on"}
        self.write_run_times = False
        self.write_run_times_args = {
            "help": "Write back probabilities after each test run",
            "action": "store_true",
        }
        self.unseed_check_ratio: float = 0.05
        self.unseed_check_ratio_args = {
            "help": "Probability for doing determinism check"
        }
        self.test_source_dir: Path = Path("tests")
        self.test_source_dir_args = {
            "type": Path,
            "help": "Root directory containing test type subdirectories (e.g., slow, fast) which hold .toml test files.",
            "env_name": "JOSHUA_TEST_FILES_DIR",
        }
        self.test_types_to_run: List[str] = ["slow", "fast", "restarting", "rare", "noSim"]
        self.test_types_to_run_args: dict = {
            "nargs": "*",
            "help": "List of test type subdirectories (under test_source_dir) to run tests from (e.g., slow, fast)."
        }
        self.trace_format: str = "json"
        self.trace_format_args = {
            "choices": ["json", "xml"],
            "help": "What format fdb should produce",
        }
        self.crash_on_error: bool = True
        self.crash_on_error_args = {
            "long_name": "no_crash",
            "action": "store_false",
            "help": "Don't crash on first error",
        }
        self.max_warnings: int = 10
        self.max_warnings_args = {"short_name": "W"}
        self.max_errors: int = 10
        self.max_errors_args = {"short_name": "E"}
        self.old_binaries_path: Path = Path("/app/deploy/global_data/oldBinaries/")
        self.old_binaries_path_args = {
            "help": "Path to the directory containing the old fdb binaries"
        }
        self.tls_plugin_path: Path = Path("/app/deploy/runtime/.tls_5_1/FDBLibTLS.so")
        self.tls_plugin_path_args = {
            "help": "Path to the tls plugin used for binaries < 5.2.0"
        }
        self.disable_kaio: bool = False
        self.use_valgrind: bool = False
        self.use_valgrind_args = {"action": "store_true"}
        self.buggify = BuggifyOption("random")
        self.buggify_args = {"short_name": "b", "choices": ["on", "off", "random"]}
        self.pretty_print: bool = False
        self.pretty_print_args = {"short_name": "P", "action": "store_true"}
        self.clean_up: bool = True
        self.clean_up_args = {"long_name": "no_clean_up", "action": "store_false"}
        self.run_temp_dir: Path | None = None
        self.run_temp_dir_args = {
            "type": Path,
            "help": "Temporary directory for individual test run artifacts and logs.",
            "required": True,
            "env_name": "TH_RUN_TEMP_DIR",
        }
        self.joshua_seed: int = random.randint(0, 2**32 - 1)
        self.joshua_seed_args = {
            "short_name": "s",
            "help": "A random seed",
            "env_name": "JOSHUA_SEED",
        }
        self.no_verbose_on_failure: bool = False
        self.no_verbose_on_failure_args = {
            "action": "store_true",
            "help": "Do not dump all trace events to summary on test failure.",
        }
        self.print_coverage = False
        self.print_coverage_args = {"action": "store_true"}
        self.binary = Path("bin") / (
            "fdbserver.exe" if os.name == "nt" else "fdbserver"
        )
        self.binary_args = {"help": "Path to executable"}
        self.fdbserver_memory: str | None = None
        self.fdbserver_memory_args = {
            "type": str,
            "required": False,
            "help": "Pass --memory SIZE to fdbserver (for example 12288MiB)",
        }
        self.hit_per_runs_ratio: int = 20000
        self.hit_per_runs_ratio_args = {
            "help": "Maximum test runs before each code probe hit at least once"
        }
        self.code_probe_reporting_sample_rate: float = 1.0
        self.code_probe_reporting_sample_rate_args = {
            "help": "Probability that a test run reports code probes"
        }
        self.output_format: str = "xml"
        self.output_format_args = {
            "short_name": "O",
            "choices": ["json", "xml"],
            "help": "What format TestHarness should produce",
        }
        self.include_test_files: str = r".*"
        self.include_test_files_args = {
            "help": "Only consider test files whose path match against the given regex"
        }
        self.exclude_test_files: str = r".^"
        self.exclude_test_files_args = {
            "help": "Don't consider test files whose path match against the given regex"
        }
        self.include_test_classes: str = r".*"
        self.include_test_classes_args = {
            "help": "Only consider tests whose names match against the given regex"
        }
        self.exclude_test_names: str = r".^"
        self.exclude_test_names_args = {
            "help": "Don't consider tests whose names match against the given regex"
        }
        self.details: bool = False
        self.details_args = {
            "help": "Print detailed results",
            "short_name": "c",
            "action": "store_true",
        }
        self.success: bool = False
        self.success_args = {"help": "Print successful results", "action": "store_true"}
        self.cov_include_files: str = r".*"
        self.cov_include_files_args = {
            "help": "Only consider coverage traces that originated in files matching regex"
        }
        self.cov_exclude_files: str = r".^"
        self.cov_exclude_files_args = {
            "help": "Ignore coverage traces that originated in files matching regex"
        }
        self.max_stderr_bytes: int = 10000
        self.write_stats: bool = True
        self.read_stats: bool = True
        self.reproduce_prefix: str | None = None
        self.reproduce_prefix_args = {
            "type": str,
            "required": False,
            "help": "When printing the results, prepend this string to the command",
        }
        self.long_running: bool = False
        self.long_running_args = {"action": "store_true"}
        self.archive_logs_on_failure: bool = False
        self.archive_logs_on_failure_args = {
            "action": "store_true",
            "help": "If set, archive FDB logs and test harness outputs to a .tar.gz file in the joshua_output_dir on test failure.",
            "env_name": "TH_ARCHIVE_LOGS_ON_FAILURE",
        }
        self.test_args_file: Path | None = None
        self.test_args_file_args = {
            "type": Path,
            "required": False,
            "help": "File containing test arguments (e.g., '-f tests/fast/CycleTest.toml -s 315315 -b off --reseed-time 100')",
        }

        self._env_names: Dict[str, str] = {}
        self._config_map = self._build_map()
        self._read_env()
        self.random.seed(self.joshua_seed, version=2)

    def change_default(self, attr: str, default_val):
        assert attr in self._config_map, "Unknown config attribute {}".format(attr)
        self.__setattr__(attr, default_val)
        self._config_map[attr].kwargs["default"] = default_val

    def _get_env_name(self, var_name: str) -> str:
        return self._env_names.get(var_name, "TH_{}".format(var_name.upper()))

    def dump(self):
        for attr in dir(self):
            obj = getattr(self, attr)
            if (
                attr == "random"
                or attr.startswith("_")
                or callable(obj)
                or attr.endswith("_args")
            ):
                continue
            print("config.{}: {} = {}".format(attr, type(obj), obj))

    def _build_map(self) -> OrderedDict[str, ConfigValue]:
        config_map: OrderedDict[str, ConfigValue] = collections.OrderedDict()
        for attr in dir(self):
            obj = getattr(self, attr)
            if attr == "random" or attr.startswith("_") or callable(obj):
                continue
            if attr.endswith("_args"):
                name = attr[0 : -len("_args")]
                assert name in config_map
                assert isinstance(obj, dict)
                for k, v in obj.items():
                    if k == "env_name":
                        self._env_names[name] = v
                    else:
                        config_map[name].kwargs[k] = v
            else:
                # attribute_args has to be declared after the attribute
                assert attr not in config_map
                val_type = type(obj)
                kwargs = {"type": val_type, "default": obj}
                config_map[attr] = ConfigValue(attr, **kwargs)
        return config_map

    def _read_env(self):
        for attr in dir(self):
            obj = getattr(self, attr)
            if (
                attr == "random"
                or attr.startswith("_")
                or attr.endswith("_args")
                or callable(obj)
            ):
                continue
            env_name = self._get_env_name(attr)
            attr_type = self._config_map[attr].kwargs["type"]
            assert attr_type is not type(None)
            e = os.getenv(env_name)
            if e is not None:
                # Use the env var to supply the default value, so that if the
                # environment variable is set and the corresponding command line
                # flag is not, the environment variable has an effect.
                self._config_map[attr].kwargs["default"] = attr_type(e)

    def build_arguments(self, parser: argparse.ArgumentParser):
        for val in self._config_map.values():
            val.add_to_args(parser)

    def extract_args(self, args: argparse.Namespace):
        for val in self._config_map.values():
            k, v = val.get_value(args)
            if v is not None:
                config.__setattr__(k, v)
        self.random.seed(self.joshua_seed, version=2)


config = Config()

if __name__ == "__main__":
    # test the config setup
    parser = argparse.ArgumentParser(
        "TestHarness Config Tester",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    config.build_arguments(parser)
    args = parser.parse_args()
    config.extract_args(args)
    config.dump()
